### PR TITLE
feat: Enable profile diff between orgs

### PR DESCRIPTION
**Pull request by genoud**
Profile project to Org
Profile diff source org to target org
Issue #243

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -929,6 +930,36 @@ EXAMPLE
 
 _See code: [src\commands\sfpowerkit\org\orgcoverage.ts](https://github.com/Accenture/sfpowerkit/blob/master/src/commands/sfpowerkit/org/orgcoverage.ts)_
 
+## `sfpowerkit:org:profile:diff`
+
+Compare profiles from project again target org or between two orgs (source and target).
```
Comment (azlam-abdulsalam) [inline README.md:935]: Spelling

File: README.md [hunk 2]
```diff
@@ -929,6 +930,36 @@ EXAMPLE
...
+
+```
+USAGE
+  $ sfdx sfpowerkit:org:profile:diff [-p <array>] [-s <string>] [-d <string>] [-u <string>] [--apiversion <string>] [--json] [--loglevel
+  trace|debug|info|warn|error|fatal|TRACE|DEBUG|INFO|WARN|ERROR|FATAL]
+
+OPTIONS
+  -d, --output=output                                                               Output folder. PRovide the output folder if comparing profiles from source
+                                                                                    org.
+  -p, --profilelist=profilelist                                                     List of profiles to compare. If not provided and no sourceusername is
```
Comment (azlam-abdulsalam) [inline README.md:945]: Is it going to be Coma Seperated ? it would be worth calling it out

File: README.md [hunk 3]
```diff
@@ -929,6 +930,36 @@ EXAMPLE
...
+                                                                                    provided, all profiles from the source folder will be processed.
+  -s, --sourceusername=sourceusername                                               Source org. If no profile is provided in the profilelist parameter, all the
+                                                                                    profile from this org will be fetched
+  -u, --targetusername=targetusername                                               username or alias for the target org; overrides default target org
+  --apiversion=apiversion                                                           override the api version used for api requests made by this command
+  --json                                                                            format output as json
+  --loglevel=(trace|debug|info|warn|error|fatal|TRACE|DEBUG|INFO|WARN|ERROR|FATAL)  [default: warn] logging level for this command invocation
+
+EXAMPLE
+  $ sfdx sfpowerkit:org:profile:diff --profilelist profilenames --targetusername username (Compare liste profiles path again target org)
```
Comment (azlam-abdulsalam) [inline README.md:955]: Spellings again

File: src/commands/sfpowerkit/org/profile/diff.ts [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    },
+    display() {
+      if (Array.isArray(this.data) && this.data.length) {
+        this.ux.table(this.data, this.tableColumnData);
+      }
+    }
+  };
+
+  protected static requiresUsername = true;
+  protected static requiresProject = true;
```
Comment (azlam-abdulsalam) [inline src/commands/sfpowerkit/org/profile/diff.ts:81]: Do We really need to worry about this?

File: src/impl/source/profiles/profileDiff.ts [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+        isDevHub: false
+      });
+    }
+    if (
+      (!this.profileList || this.profileList.length === 0) &&
+      this.sourceOrgStr
+    ) {
+      this.sourceLabel = this.sourceOrg.getConnection().getUsername();
+      SFPowerkit.log(
+        "Not profile provided, looding profile from source org. ",
```
Comment (azlam-abdulsalam) [inline src/impl/source/profiles/profileDiff.ts:50]: Spelling and let us update to loading all profiles from source org

File: src/impl/source/profiles/profileDiff.ts [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+        [{ type: "Profile", folder: null }],
+        conn
+      );
+      profileSource = profileNamesPromise.then(profileNames => {
+        return this.retrieveProfiles(profileNames, this.sourceOrg);
+      });
+    } else {
+      SFPowerkit.log("Reading profiles from file system. ", LoggerLevel.INFO);
+      if (!this.profileList || this.profileList.length === 0) {
+        let srcFolders = await SFPowerkit.getProjectDirectories();
```
Comment (azlam-abdulsalam) [inline src/impl/source/profiles/profileDiff.ts:65]: Can we have a specific folder to look into flag?

**Comment by azlam-abdulsalam** [comment]
Also what's the behaviour if source username is provided? Does it imply we overwrite any profile that are existing in the project directory with diff markers along with new one, Should we move that to a temporary directory?

File: src/commands/sfpowerkit/org/profile/diff.ts (hunk 4 above)
Comment (genoud) [reply src/commands/sfpowerkit/org/profile/diff.ts:81]: Yes, can make the command run outside of a project folder. For now when no profile is provided, the project file is required to load all profiles in the project.

**Comment by genoud** [comment]
> Also what's the behaviour if source username is provided? Does it imply we overwrite any profile that are existing in the project directory with diff markers along with new one, Should we move that to a temporary directory?

The source username is provided, the user need to provide an output folder where the profiles with diff marker will be stored.

File: README.md (hunk 1 above)
Comment (manisfdcsfdx) [reply README.md]: resolved

File: README.md (hunk 2 above)
Comment (manisfdcsfdx) [reply README.md]: resolved

File: README.md (hunk 3 above)
Comment (manisfdcsfdx) [reply README.md]: resolved

File: src/impl/source/profiles/profileDiff.ts (hunk 5 above)
Comment (manisfdcsfdx) [reply src/impl/source/profiles/profileDiff.ts]: resolved

**Comment by manisfdcsfdx** [comment]
@genoud command hasnt worked for me yet. tried in both mac and windows

all profiles in project
````
AMAC02Z10EHLVDQ:profiletest ezdhan.hussain.s.k$ sfdx sfpowerkit:org:profile:diff -d profilediff -u buildbot
INFO  [2020-04-28 10:29:45.070 +0000] (sfpowerkit): Profile diff start.
INFO  [2020-04-28 10:29:45.076 +0000] (sfpowerkit): Reading profiles from file system.
Reading from File System PROGRESS  | ████████████████████████████████████████ | 11/11 Profiles
Retrieving From buildbot@example.com PROGRESS  | ████████████████████████████████████████ | 11/11 Profiles
INFO  [2020-04-28 10:32:01.266 +0000] (sfpowerkit): Handling diff
Diff processing PROGRESS  | ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ | 0/11 Profiles
````

trying for admin profile
````
AMAC02Z10EHLVDQ:profiletest ezdhan.hussain.s.k$ sfdx sfpowerkit:org:profile:diff -d profilediff -u buildbot -p Admin
INFO  [2020-04-28 10:44:20.633 +0000] (sfpowerkit): Profile diff start.
INFO  [2020-04-28 10:44:20.637 +0000] (sfpowerkit): Reading profiles from file system.
Reading from File System PROGRESS  | ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ | 0/1 ProfilesERROR running sfpowerkit:org:profile:diff:  ENOENT: no such file or directory, open 'Admin'
````

**Comment by genoud** [comment]
> @genoud command hasnt worked for me yet. tried in both mac and windows
> 
> all profiles in project - just hangs like this for more than 30 mins
> 
> ```
> AMAC02Z10EHLVDQ:profiletest ezdhan.hussain.s.k$ sfdx sfpowerkit:org:profile:diff -d profilediff -u buildbot
> INFO  [2020-04-28 10:29:45.070 +0000] (sfpowerkit): Profile diff start.
> INFO  [2020-04-28 10:29:45.076 +0000] (sfpowerkit): Reading profiles from file system.
> Reading from File System PROGRESS  | ████████████████████████████████████████ | 11/11 Profiles
> Retrieving From buildbot@example.com PROGRESS  | ████████████████████████████████████████ | 11/11 Profiles
> INFO  [2020-04-28 10:32:01.266 +0000] (sfpowerkit): Handling diff
> Diff processing PROGRESS  | ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ | 0/11 Profiles
> ```
> 
> trying for admin profile
> 
> ```
> AMAC02Z10EHLVDQ:profiletest ezdhan.hussain.s.k$ sfdx sfpowerkit:org:profile:diff -d profilediff -u buildbot -p Admin
> INFO  [2020-04-28 10:44:20.633 +0000] (sfpowerkit): Profile diff start.
> INFO  [2020-04-28 10:44:20.637 +0000] (sfpowerkit): Reading profiles from file system.
> Reading from File System PROGRESS  | ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ | 0/1 ProfilesERROR running sfpowerkit:org:profile:diff:  ENOENT: no such file or directory, open 'Admin'
> ```

Hello @manisfdcsfdx 
I was expecting this issue with big profile files.
With my developer edition where profile are around 10k lines, I was getting a really bad experience. (1-2 minute per profile). This is due to the `jsdiff `library that compute the diff. It does not perform well on big content.

You can run the command with debug flag to see where it get stuck.

We need to find a different way of doing that, otherwise it will not be usable.

**Comment by azlam-abdulsalam** [comment]
@genoud What if we give the notification to the user each profile by profile, so the feeling of not stuck is there, also lets look at a more performant dff

**Comment by genoud** [comment]
Yes @azlam-abdulsalam , the progress is updated by profile already.
Yes, I'm looking into alternatives to this library.

**Comment by manisfdcsfdx** [comment]
lets target this for next release [1.45.x](https://github.com/Accenture/sfpowerkit/milestone/18). i will descope it from this week release

**Comment by genoud** [comment]
@manisfdcsfdx  can you test again?

**Comment by manisfdcsfdx** [comment]
@genoud when i retrieve the latest code am getting few compilation errors
image

**Comment by azlam-abdulsalam** [comment]
@manisfdcsfdx CI seems okie

**Comment by genoud** [comment]
> @genoud when i retrieve the latest code am getting few compilation errors
> image

@manisfdcsfdx  still having issue. Not having any issue on my end. Can you try reverting changes on `npm-shrinkwrap.json` file and retry?

**Comment by manisfdcsfdx** [comment]
@genoud seems like my machine issue. i restarted my machine and vsc then problem. 

the command works like a charm and a awesome feature. performance is also awesome.

**Comment by azlam-abdulsalam** [comment]
> Approved from my side, if Azlam is okay with code then we can merge.

Happy to merge once we remove the progress bar in impl

**Comment by manisfdcsfdx** [comment]
@genoud we need progress bar fix here as well. 

@azlam-abdulsalam FYI

**Comment by genoud** [comment]
> @genoud we need progress bar fix here as well.
> 
> @azlam-abdulsalam FYI

Done!

**Pull request merged**